common/nym-kcp/src/session.rs

```diff
@@ -0,0 +1,1778 @@
+        self.snd_una = cmp::max(self.snd_una, una);
+    }
+
+    fn parse_ack(&mut self, sn: u32, ts: u32) {
+        debug!(
+            "[ConvID: {}, Thread: {:?}] Parsing ack, sn: {}, ts: {}",
+            self.conv,
+            thread::current().id(),
+            sn,
+            ts
+        );
+        // find the segment in snd_buf
+        if let Some(pos) = self.snd_buf.iter().position(|seg| seg.sn == sn) {
+            let seg = self.snd_buf.remove(pos).unwrap();
+            debug!(
+                "[ConvID: {}, Thread: {:?}] Acked segment, sn: {}, frg: {}",
+                self.conv,
+                thread::current().id(),
+                sn,
+                seg.frg
+            );
+            // update RTT
+            let rtt = self.current.saturating_sub(ts);
+            self.update_rtt(rtt);
+        } else {
+            // Log if the segment was NOT found
+            let current_sns: Vec<u32> = self.snd_buf.iter().map(|s| s.sn).collect();
+            warn!(
+                "[ConvID: {}, Thread: {:?}] parse_ack: ACK received for sn={}, but segment not found in snd_buf (len={}): {:?}",
+                self.conv,
+                thread::current().id(),
+                sn,
+                self.snd_buf.len(),
+                current_sns
+            );
+        }
+    }
+
+    fn parse_data(&mut self, pkt: &KcpPacket) {
+        // Insert into rcv_buf if pkt.sn in [rcv_nxt .. rcv_nxt + rcv_wnd)
+        if pkt.sn() >= self.rcv_nxt + self.rcv_wnd as u32 {
+            // out of window
+            return;
+        }
+        if pkt.sn() < self.rcv_nxt {
+            // already got it, discard
+            return;
+        }
+
+        // Check if we have it
+        let mut insert_idx = self.rcv_buf.len();
+        for (i, seg) in self.rcv_buf.iter().enumerate() {
+            #[allow(clippy::comparison_chain)]
+            if pkt.sn() < seg.sn {
+                insert_idx = i;
+                break;
+            } else if pkt.sn() == seg.sn {
+                // duplicate
+                return;
+            }
+        }
+
+        let seg = Segment {
+            sn: pkt.sn(),
+            frg: pkt.frg(),
+            ts: pkt.ts(),
+            resendts: 0,
+            rto: 0,
+            xmit: 0,
+            data: pkt.data().into(),
+        };
+        self.rcv_buf.insert(insert_idx, seg);
+
+        // Move ready segments from rcv_buf -> rcv_queue
+        self.move_buf_to_queue();
+    }
+
+    fn move_buf_to_queue(&mut self) {
+        // Loop as long as we can potentially extract a complete message from the front
+        loop {
+            // Check if the buffer starts with the next expected sequence number
+            if self.rcv_buf.is_empty() || self.rcv_buf[0].sn != self.rcv_nxt {
+                break; // Cannot start assembling a message now
+            }
+
+            // Scan ahead in rcv_buf to find if a complete message exists contiguously
```

Comment:
why are we concerned in delivering full messages as opposed to next fragment in order? should it be KCP's concern to de-fragment data?